src/main/java/net/neoforged/neoforge/common/crafting/BlockTagIngredient.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.common.crafting;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import net.minecraft.core.Holder;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.core.registries.Registries;
+import net.minecraft.network.chat.Component;
+import net.minecraft.tags.TagKey;
+import net.minecraft.world.item.ItemStack;
+import net.minecraft.world.item.crafting.Ingredient;
+import net.minecraft.world.level.block.Block;
+import net.minecraft.world.level.block.Blocks;
+import net.neoforged.neoforge.common.NeoForgeMod;
+
+/**
+ * Ingredient that matches ItemStacks of blocks from a TagKey<Block>, useful in cases
+ * like "minecraft:convertable_to_mud" where there isn't an accompanying item tag
+ * <p>
+ * Notice: This should not be used as a replacement for the normal Item Tag Ingredient,
+ * This should only be used when there is no way an item tag can be used in your use case
+ */
+public class BlockTagIngredient extends Ingredient {
+    public static final Codec<BlockTagIngredient> CODEC = RecordCodecBuilder.create(
+            valueInstance -> valueInstance.group(TagKey.codec(Registries.BLOCK).fieldOf("tag").forGetter(BlockTagIngredient::getTag))
+                    .apply(valueInstance, BlockTagIngredient::new));
```

Comment:
Why is this different from the 1.20.6 PR, which has this as a `MapCodec`?

Reply:
for 1.20.4 it needs to be a Codec instead of a MapCodec